### PR TITLE
Lm/watch

### PR DESCRIPTION
**Pull request by leemaguire**

### CODE REVIEW

File: Realm/RLMSyncManager.mm [hunk 1]
```diff
@@ -150,8 +150,6 @@ - (void)configureWithRootDirectory:(NSURL *)rootDirectory appConfiguration:(RLMA
 
     if (appConfiguration) {
         SyncManager::shared().configure(config, [appConfiguration config]);
-    } else {
```
Comment (jsflax) [inline Realm/RLMSyncManager.mm:153]: Why was this removed?

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 2]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+
+- (void)performWatchTest:(nullable dispatch_queue_t)delegateQueue {
+    XCTestExpectation *expectation = [self expectationWithDescription:@"watch collection and receive change event 5 times"];
+
+    RLMMongoClient *client = [self.anonymousUser mongoClientWithServiceName:@"mongodb1"];
+    RLMMongoDatabase *database = [client databaseWithName:@"test_data"];
+    __block RLMMongoCollection *collection = [database collectionWithName:@"Dog"];
+
+    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (int i=0; i<5; i++) {
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2265]: Weird spacing.

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 3]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+            [collection insertOneDocument:@{@"name": @"fido"} completion:^(RLMObjectId * objectId, NSError * error) {
+                XCTAssertNil(error);
+                XCTAssertNotNil(objectId);
+            }];
+            [NSThread sleepForTimeInterval:2];
+            if (i==4) {
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2271]: weird spacing

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 4]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+                                                                                  completion:^(NSError * error) {
+        XCTAssertNil(error);
+        [expectation fulfill];
+    }];
+
+    [collection watchWithDelegate:testUtility delegateQueue:delegateQueue];
+    [self waitForExpectations:@[expectation] timeout:60.0];
+}
+
+- (void)testWatchWithMatchFilter {
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2288]: What is the point of this method?

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 5]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+    [self performWatchWithMatchFilterTest:nil];
+}
+
+- (void)testWatchWithMatchFilterAsync {
+    auto asyncQueue = dispatch_queue_create("io.realm.watchQueue", DISPATCH_QUEUE_CONCURRENT);
+    [self performWatchWithMatchFilterTest:asyncQueue];
+}
+
+- (void)performWatchWithMatchFilterTest:(nullable dispatch_queue_t)delegateQueue {
+
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2298]: extra newline

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 6]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+                                                                                  completion:^(NSError * error) {
+        XCTAssertNil(error);
+        [expectation fulfill];
+    }];
+    [collection watchWithMatchFilter:@{@"fullDocument._id": objectIds[0]}
+                            delegate:testUtility
+                       delegateQueue:delegateQueue];
+
+    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (int i=0; i<5; i++) {
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2330]: weird spacing

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 7]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+                XCTAssertNil(error);
+            }];
+
+            [collection updateOneDocumentWhere:@{@"_id": objectIds[1]}
+                                updateDocument:@{@"breed": @"french bulldog", @"name": [NSString stringWithFormat:@"fido-%d", i]}
+                                    completion:^(RLMUpdateResult * _Nullable, NSError * error) {
+                XCTAssertNil(error);
+            }];
+            [NSThread sleepForTimeInterval:2];
+            if (i==4) {
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2343]: not going to keep marking these but use proper spacing please

File: Realm/RLMMongoCollection.h [hunk 8]
```diff
@@ -18,12 +18,27 @@
 
 #import <Foundation/Foundation.h>
 #import <Realm/RLMObjectId.h>
+#import <Realm/RLMNetworkTransport.h>
```
Comment (jsflax) [inline Realm/RLMMongoCollection.h:21]: Any way we can avoid this import here?

File: Realm/RLMMongoCollection.h [hunk 9]
```diff
@@ -18,12 +18,27 @@
...
 
 NS_ASSUME_NONNULL_BEGIN
 @protocol RLMBSON;
 
 @class RLMApp, RLMFindOptions, RLMFindOneAndModifyOptions, RLMUpdateResult;
 
+/// Delegate which is used for subscribing to changes a  `watch` stream.
+@protocol RLMChangeEventDelegate
+/// The stream was opened.
+- (void)didOpen;
```
Comment (jsflax) [inline Realm/RLMMongoCollection.h:31]: These should be named more explicitly: `changeStreamDidOpen:(RLMChangeStream)` `changeStream:(RLMChangeStream) didCloseWithError:(nullable NSError *)` etc. Whatever `RLMChangeStream` is should have some basic closing functionality.

File: Realm/ObjectServerTests/RLMObjectServerTests.mm (hunk 4 above)
Comment (leemaguire) [reply Realm/ObjectServerTests/RLMObjectServerTests.mm:2288]: To test `watchWithMatchFilter:` on the main thread

File: Realm/RLMMongoCollection.h (hunk 8 above)
Comment (leemaguire) [reply Realm/RLMMongoCollection.h:21]: No because we need `RLMEventDelegate`

Comment (jsflax) [reply Realm/RLMMongoCollection.h:21]: Can't we forward declare it?

Comment (leemaguire) [reply Realm/RLMMongoCollection.h:21]: Forward declaring it gives a build error basically stating the delegate type cannot be found. Same goes for trying to forward declare the protocol.

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,7 @@ x.y.z Release notes (yyyy-MM-dd)
 =============================================================
 ### Enhancements
 * Add support for the 64-bit watchOS simulator added in Xcode 12.
+* Add ability to stream change events on a remote MongoDB collection with `[RLMMongoCollection watch:delegate:delegateQueue]`, `MongoCollection.watch(delegate)`. When calling `watch(delegate)` you will be given a `RLMChangeStream` (`ChangeStream`), this will be used to invalidate and stop the streaming session by calling `[RLMChangeStream close]` (`ChangeStream,close()`) when needed.
```
Comment (tgoyne) [inline CHANGELOG.md:5]:
```suggestion
* Add ability to stream change events on a remote MongoDB collection with `[RLMMongoCollection watch:delegate:delegateQueue]`, `MongoCollection.watch(delegate)`. When calling `watch(delegate)` you will be given a `RLMChangeStream` (`ChangeStream`), this will be used to invalidate and stop the streaming session by calling `[RLMChangeStream close]` (`ChangeStream.close()`) when needed.
```

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 11]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+
+    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (int i = 0; i < 5; i++) {
+            [collection insertOneDocument:@{@"name": @"fido"} completion:^(RLMObjectId * objectId, NSError * error) {
+                XCTAssertNil(error);
+                XCTAssertNotNil(objectId);
+            }];
+            [NSThread sleepForTimeInterval:2];
+            if (i == 4) {
+                [NSThread sleepForTimeInterval:10];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2279]: Why does this need to sleep? Can't it just close the changestream on the main thread after waiting for expectations?

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 12]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+    }];
+    __block RLMChangeStream *changeStream = [collection watchWithDelegate:testUtility delegateQueue:delegateQueue];
+
+    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (int i = 0; i < 5; i++) {
+            [collection insertOneDocument:@{@"name": @"fido"} completion:^(RLMObjectId * objectId, NSError * error) {
+                XCTAssertNil(error);
+                XCTAssertNotNil(objectId);
+            }];
+            [NSThread sleepForTimeInterval:2];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2277]: Is this sleep just to prevent change notifications from being batched? If so it'd be better to just wait for each change event to be delivered before doing the next insert, which would be more robust and probably faster in practice.

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 13]
```diff
@@ -2242,4 +2243,254 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+                                    completion:^(RLMUpdateResult * _Nullable, NSError * error) {
+                XCTAssertNil(error);
+            }];
+
+            [collection updateOneDocumentWhere:@{@"_id": objectIds[1]}
+                                updateDocument:@{@"breed": @"french bulldog", @"name": [NSString stringWithFormat:@"fido-%d", i]}
+                                    completion:^(RLMUpdateResult * _Nullable, NSError * error) {
+                XCTAssertNil(error);
+            }];
+            [NSThread sleepForTimeInterval:2];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2341]: Same comments about sleep as above; using sleep for synchronization either makes the tests take forever to run or causes spurious failures (or both).

File: Realm/ObjectServerTests/SwiftObjectServerTests.swift [hunk 14]
```diff
@@ -1531,4 +1531,442 @@ class SwiftObjectServerTests: SwiftSyncTestCase {
...
+
+        DispatchQueue.global().async {
+            for i in 0..<5 {
+                collection.insertOne(document) { (_, error) in
+                    XCTAssertNil(error)
+                }
+                Thread.sleep(forTimeInterval: 2)
+                if i == 4 {
+                    Thread.sleep(forTimeInterval: 10)
+                    subscriptions.forEach {$0.cancel()}
```
Comment (tgoyne) [inline Realm/ObjectServerTests/SwiftObjectServerTests.swift:1814]: Spaces inside {} for swift closures

File: Realm/ObjectServerTests/RLMWatchTestUtility.m [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    if (error) {
+        return _completion(error);
+    }
+
+    if ((_currentChangeEventCount == _targetChangeEventCount) && _didOpenWasCalled) {
+        return _completion(nil);
+    }
+}
+
+- (void)changeStreamDidOpen:(nonnull RLMChangeStream *)changeStream {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMWatchTestUtility.m:62]:
```suggestion
- (void)changeStreamDidOpen:(nonnull __unused RLMChangeStream *)changeStream {
```

File: Realm/RLMMongoCollection.h (hunk 8 above)
Comment (tgoyne) [reply Realm/RLMMongoCollection.h:21]: Declaring a conformance to a protocol is one of the things that needs the full definition.

File: Realm/RLMApp.mm [hunk 16]
```diff
@@ -18,6 +18,7 @@
 
 #import "RLMApp_Private.hpp"
 
+#import "RLMNetworkTransport_Private.hpp"
```
Comment (tgoyne) [inline Realm/RLMApp.mm:21]: This appears to be unused?

File: Realm/RLMMongoCollection.mm [hunk 17]
```diff
@@ -28,6 +28,60 @@
...
 #import "sync/remote_mongo_collection.hpp"
 
+@implementation RLMChangeStream {
+    realm::app::WatchStream _watchStream;
+    id<RLMChangeEventDelegate> _subscriber;
+    __weak NSURLSession *_session;
+}
+- (instancetype)initWithChangeEventSubscriber:(id<RLMChangeEventDelegate>)subscriber {
+    if (self = [super init]) {
+        _watchStream = realm::app::WatchStream();
```
Comment (tgoyne) [inline Realm/RLMMongoCollection.mm:38]: This is redundant; the ivar is already default constructed when we get here.

File: Realm/RLMMongoCollection.mm [hunk 18]
```diff
@@ -28,6 +28,60 @@
...
+- (void)didOpen {
+    [_subscriber changeStreamDidOpen:self];
+}
+
+- (void)didReceiveError:(nonnull NSError *)error {
+    [_subscriber changeStreamDidReceiveError:error];
+}
+
+- (void)didReceiveEvent:(nonnull NSData *)event {
+    auto str = [[NSString alloc] initWithData:event encoding:NSUTF8StringEncoding].UTF8String;
```
Comment (tgoyne) [inline Realm/RLMMongoCollection.mm:58]: This is just `event.data`?

File: Realm/RLMNetworkTransport.mm [hunk 19]
```diff
@@ -85,6 +89,40 @@ - (void)sendRequestToServer:(RLMRequest *) request
...
+    [[session dataTaskWithURL:url] resume];
+    [subscriber didOpen];
+    return session;
+}
+
+- (RLMRequest *)RLMRequestFromRequest:(realm::app::Request)request {
+    RLMRequest *rlmRequest = [RLMRequest new];
+    NSMutableDictionary<NSString *, NSString*> *headersDict = [NSMutableDictionary new];
+    for(auto &header : request.headers) {
+        [headersDict setValue:@(header.second.c_str()) forKey:@(header.first.c_str())];
```
Comment (tgoyne) [inline Realm/RLMNetworkTransport.mm:116]:
```suggestion
    for (auto& [key, value] : request.headers) {
        headersDict[@(key.c_str())] = @(value.c_str());
```

File: Realm/RLMNetworkTransport.mm [hunk 20]
```diff
@@ -85,6 +89,40 @@ - (void)sendRequestToServer:(RLMRequest *) request
...
+    auto session = [NSURLSession sessionWithConfiguration:sessionConfig
+                                                 delegate:delegate
+                                            delegateQueue:nil];
+    NSURL *url = [[NSURL alloc] initWithString:request.url];
+    [[session dataTaskWithURL:url] resume];
+    [subscriber didOpen];
+    return session;
+}
+
+- (RLMRequest *)RLMRequestFromRequest:(realm::app::Request)request {
```
Comment (tgoyne) [inline Realm/RLMNetworkTransport.mm:112]: This should take a const ref to avoid copying the request an extra time.

File: Realm/RLMNetworkTransport_Private.hpp [hunk 21]
```diff
@@ -0,0 +1,23 @@
+#import "RLMNetworkTransport.h"
```
Comment (tgoyne) [inline Realm/RLMNetworkTransport_Private.hpp:1]: Missing license header

File: RealmSwift/MongoClient.swift [hunk 22]
```diff
@@ -483,4 +502,230 @@ extension MongoCollection {
...
+        init(collection: MongoCollection, queue: DispatchQueue, filterIds: [ObjectId]? = nil, matchFilter: Document? = nil) {
+            self.collection = collection
+            self.queue = queue
+            self.filterIds = filterIds
+            self.matchFilter = matchFilter
+        }
+
+        /// :nodoc:
+        public func receive<S: Subscriber>(subscriber: S) where
+            Self.Failure == S.Failure, Self.Output == S.Input {
```
Comment (tgoyne) [inline RealmSwift/MongoClient.swift:668]:
```suggestion
            Failure == S.Failure, Output == S.Input {
```

`Self.` isn't applicable here since it isn't a protocol.

File: RealmSwift/MongoClient.swift [hunk 23]
```diff
@@ -483,4 +502,230 @@ extension MongoCollection {
...
+@available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, *)
+extension Publishers {
+
+    class WatchSubscription<S: Subscriber>: ChangeEventDelegate, Subscription where S.Input == AnyBSON, S.Failure == Error {
+
+        private let collection: MongoCollection
+        private var changeStream: ChangeStream?
+        private var subscriber: S?
+
+        init(collection: MongoCollection, subscriber: S, queue: DispatchQueue = .main) {
```
Comment (tgoyne) [inline RealmSwift/MongoClient.swift:592]: This probably just wants to be one initializer that has some optional parameters? Checking which parameters were actually supplied in WatchPublisher rather than here is just adding some extra boilerplate.

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 24]
```diff
@@ -2242,4 +2243,286 @@ - (void)testAddPropertyToReadOnlyRealmWithAsyncOpen {
...
+    }];
+    [self waitForExpectations:@[insertManyExpectation] timeout:60.0];
+
+    XCTestExpectation *expectation = [self expectationWithDescription:@"watch collection and receive change event 3 times"];
+
+    RLMWatchTestUtility *testUtility =
+    [[RLMWatchTestUtility alloc] initWithChangeEventCount:3
+                                         matchingObjectId:objectIds[0]
+                                            eventReceived:^(NSError * error) {
+        XCTAssertNil(error);
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:2397]: It looks like every use of RLMWatchTestUtility has the exact same logic in the eventReceived: and completion: blocks, so that could just be pushed into RLMWatchTestUtility to eliminate some duplication between tests. It'd probably make sense to have the initializer take the XCTest as a parameter and provide the semaphore and expectation as properties.

File: Realm/ObjectServerTests/RLMWatchTestUtility.h [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+/// @param eventReceived Block which is invoked when a change event is received.
+/// @param completion Block which is invoked when target amount of change events has been hit, or an error has occured.
+- (instancetype)initWithChangeEventCount:(NSUInteger)changeEventCount
+                           eventReceived:(RLMWatchTestUtilityBlock)eventReceived
+                              completion:(RLMWatchTestUtilityBlock)completion;
+
+/// Sets up an object that subscribes to the RLMChangeEventDelegate
+/// @param changeEventCount The target amount of change events for the test to succeed
+/// @param matchingObjectId An objectId that the change event must match.
+/// /// @param eventReceived Block which is invoked when a change event is received.
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMWatchTestUtility.h:40]:
```suggestion
/// @param eventReceived Block which is invoked when a change event is received.
```

File: RealmSwift/MongoClient.swift [hunk 26]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+        proxyDelegate?.changeStreamDidReceive(changeEvent: bson)
+    }
+}
+
+#if canImport(Combine)
+import Combine
+
+@available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, *)
+extension Publishers {
+
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:585]: extra newline

File: RealmSwift/MongoClient.swift [hunk 27]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+    class WatchSubscription<S: Subscriber>: ChangeEventDelegate, Subscription where S.Input == AnyBSON, S.Failure == Error {
+
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:587]: extra newline

File: RealmSwift/MongoClient.swift [hunk 28]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+        private let collection: MongoCollection
+        private var changeStream: ChangeStream?
+        private var subscriber: S?
+
+        init(collection: MongoCollection, subscriber: S,
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:592]: needs newline between params

File: RealmSwift/MongoClient.swift [hunk 29]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+             queue: DispatchQueue = .main,
+             filterIds: [ObjectId]? = nil,
+             matchFilter: Document? = nil) {
+            self.collection = collection
+            self.subscriber = subscriber
+
+            if let matchFilter = matchFilter {
+                setup(with: matchFilter, queue: queue)
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:600]: These setup methods are only called here, and can just be collapsed to be one liners here.

File: RealmSwift/MongoClient.swift [hunk 30]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+
+        /// Specifies the scheduler on which to perform subscribe, cancel, and request operations.
+        ///
+        /// - parameter scheduler: The dispatch queue to perform the subscription on.
+        /// - returns: A publisher which subscribes on the given scheduler.
+        public func subscribe<S: Scheduler>(on scheduler: S) -> WatchPublisher {
+            guard let queue = scheduler as? DispatchQueue else {
+                fatalError("Cannot subscribe on scheduler \(scheduler): only dispatch queues are currently implemented.")
+            }
+            return Self(collection: collection,
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:683]: Why return `Self` here?

File: RealmSwift/MongoClient.swift [hunk 31]
```diff
@@ -483,4 +502,217 @@ extension MongoCollection {
...
+                        queue: queue,
+                        filterIds: filterIds,
+                        matchFilter: matchFilter)
+        }
+    }
+}
+
+@available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, *)
+extension MongoCollection {
+
```
Comment (jsflax) [inline RealmSwift/MongoClient.swift:693]: nó newline

**Review by jsflax — COMMENTED** [review]
Should also add a CHANGELOG entry for the publishers.

File: Realm/ObjectServerTests/RLMObjectServerTests.mm (hunk 24 above)
Comment (leemaguire) [reply Realm/ObjectServerTests/RLMObjectServerTests.mm:2397]: All logic has moved to RLMWatchTestUtility

**Review by jsflax — APPROVED** [review]
